x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/ProgressListenableActionFutureTests.java

```diff
@@ -0,0 +1,172 @@
+import org.elasticsearch.common.unit.ByteSizeUnit;
+import org.elasticsearch.common.util.concurrent.AtomicArray;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.Objects;
+import java.util.concurrent.CountDownLatch;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.hamcrest.Matchers.nullValue;
+import static org.hamcrest.Matchers.sameInstance;
+
+public class ProgressListenableActionFutureTests extends ESTestCase {
+
+    public void testOnResponseCallsListeners() {
+        final ProgressListenableActionFuture future = randomFuture();
+
+        final AtomicArray<Long> listenersResponses = new AtomicArray<>(between(0, 50));
+        for (int i = 0; i < listenersResponses.length(); i++) {
+            final int listenerIndex = i;
+            future.addListener(
+                ActionListener.wrap(
+                    progress -> listenersResponses.setOnce(listenerIndex, progress),
+                    e -> listenersResponses.setOnce(listenerIndex, null)
+                ),
+                randomLongBetween(future.start, future.end)
+            );
+        }
+        assertTrue(listenersResponses.asList().stream().allMatch(Objects::isNull));
+        future.onResponse(future.end);
+        assertTrue(listenersResponses.asList().stream().allMatch(value -> value <= future.end));
+
+        final IllegalStateException ise = expectThrows(IllegalStateException.class, () -> future.onResponse(future.end));
+        assertThat(ise.getMessage(), containsString("Future is already completed"));
+    }
+
+    public void testOnFailureCallsListeners() {
+        final ProgressListenableActionFuture future = randomFuture();
+
+        final AtomicArray<Exception> listenersResponses = new AtomicArray<>(between(0, 50));
+        for (int i = 0; i < listenersResponses.length(); i++) {
+            final int listenerIndex = i;
+            future.addListener(
+                ActionListener.wrap(
+                    o -> listenersResponses.setOnce(listenerIndex, null),
+                    e -> listenersResponses.setOnce(listenerIndex, e)
+                ),
+                randomLongBetween(future.start, future.end)
+            );
+        }
+        assertTrue(listenersResponses.asList().stream().allMatch(Objects::isNull));
+
+        final Exception exception = new ElasticsearchException("simulated");
+        future.onFailure(exception);
+
+        for (int i = 0; i < listenersResponses.length(); i++) {
+            assertThat(listenersResponses.get(i), sameInstance(exception));
+        }
+
+        IllegalStateException ise = expectThrows(IllegalStateException.class, () -> future.onFailure(exception));
+        assertThat(ise.getMessage(), containsString("Future is already completed"));
+    }
+
+    public void testProgressUpdatesCallsListeners() throws Exception {
```

Comment:
Should we also have a test that partially completes and then fails and verify that some listeners completed successfully and others failed?

Reply:
I added `testPartialProgressionThenFailure`